src/main/java/org/chocosolver/solver/search/IResolutionHelper.java

```diff
@@ -0,0 +1,387 @@
+	 *          optional criterions to stop the search before finding all/best solution
+	 * @return a list that contained the found solutions.
+	 */
+	default Stream<Solution> findAllSolutions(Criterion... stop) {
+		if (stop != null) {
+			_me().getSolver().addStopCriterion(stop);
+		}
+		Spliterator<Solution> it = new Spliterator<Solution>() {
+
+			@Override
+			public boolean tryAdvance(Consumer<? super Solution> action) {
+				if (_me().solve()) {
+					action.accept(new Solution(_me()).record());
+					return true;
+				}
+				if (stop != null) {
+					_me().getSolver().removeStopCriterion(stop);
+				}
+				return false;
+			}
+
+			@Override
+			public Spliterator<Solution> trySplit() {
+				return null;
+			}
+
+			@Override
+			public long estimateSize() {
+				return Long.MAX_VALUE;
+			}
+
+			@Override
+			public int characteristics() {
+				return Spliterator.ORDERED | Spliterator.DISTINCT | Spliterator.NONNULL | Spliterator.CONCURRENT;
+			}
+
+		};
+		return StreamSupport.stream(it, false);
+	}
+
+	/**
+	 * Attempt to find the solution that optimizes the mono-objective problem defined by a unique objective variable and
+	 * an optimization criteria.
+	 * <ul>
+	 * <li>If this method returns <i>null</i>:</li>
+	 * <ul>
+	 * <li>either the resolution stops eagerly du to a stop criterion (e.g., a time limit) and no solution has been found
+	 * so far,</li>
+	 * <li>or the problem cannot be satisfied (i.e., over constrained).</li>
+	 * </ul>
+	 * <li>If this method returns a {@link Solution}:</li>
+	 * <ul>
+	 * <li>either the resolution stops eagerly du to a stop criterion and the solution is the best found so far but there
+	 * is no guarantee that it is the optimal one,</li>
+	 * <li>or it is the optimal one.</li>
+	 * </ul>
+	 * </ul>
+	 * <p>
+	 * This method run the following instructions:
+	 *
+	 * <pre>
+	 *     {@code
+	 *     model.setObjective(maximize ? ResolutionPolicy.MAXIMIZE : ResolutionPolicy.MINIMIZE, objective);
+	 *     Solution s = new Solution(model);
+	 *     while (model.solve()) {
+	 *          s.record();
+	 *     }
+	 *     return model.getSolver().isFeasible() == ESat.TRUE ? s : null;
+	 *     }
+	 * </pre>
+	 *
+	 * @param objective
+	 *          integer variable to optimize
+	 * @param maximize
+	 *          set to <tt>true</tt> to solve a maximization problem, set to <tt>false</tt> to solve a minimization
+	 *          problem.
+	 * @param stop
+	 *          optional criterions to stop the search before finding all/best solution
+	 * @return
+	 *         <ul>
+	 *         <li><tt>null</tt> if the problem has no solution or a stop criterion stops the search before finding a
+	 *         first solution</li>
+	 *         <li>a {@link Solution} if at least one solution has been found. The solution is proven to be optimal if no
+	 *         stop criterion stops the search.</li>
+	 *         </ul>
+	 */
+	default Solution findOptimalSolution(IntVar objective, boolean maximize, Criterion... stop) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method findOptimalSolution() has an NPath complexity of 200](https://www.codacy.com/app/cprudhom/choco-solver/file/2464610826/issues/source?bid=3218989&fileBranchId=3256349#l232)